src/remote/remote_test.go

```diff
@@ -0,0 +1,271 @@
+package remote
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"testing"
+
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/bazelbuild/remote-apis/build/bazel/semver"
+	"github.com/stretchr/testify/assert"
+	bs "google.golang.org/genproto/googleapis/bytestream"
+	rpcstatus "google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/thought-machine/please/src/core"
+)
+
+func TestInit(t *testing.T) {
+	c := newClient()
+	assert.NoError(t, c.CheckInitialised())
+}
+
+func TestBadAPIVersion(t *testing.T) {
+	defer server.Reset()
+	server.HighApiVersion.Major = 1
+	server.LowApiVersion.Major = 1
+	c := newClient()
+	assert.Error(t, c.CheckInitialised())
+	assert.Contains(t, c.CheckInitialised().Error(), "1.0.0 - 1.1.0")
+}
+
+func TestUnsupportedDigest(t *testing.T) {
+	defer server.Reset()
+	server.DigestFunction = []pb.DigestFunction_Value{
+		pb.DigestFunction_SHA256,
+		pb.DigestFunction_SHA384,
+		pb.DigestFunction_SHA512,
+	}
+	c := newClient()
+	assert.Error(t, c.CheckInitialised())
+}
+
+func TestStoreAndRetrieve(t *testing.T) {
+	c := newClient()
+	c.CheckInitialised()
+	target := core.NewBuildTarget(core.BuildLabel{PackageName: "package", Name: "target1"})
+	target.AddSource(core.FileLabel{File: "src1.txt", Package: "package"})
+	target.AddSource(core.FileLabel{File: "src2.txt", Package: "package"})
+	target.AddOutput("out1.txt")
+	// The hash is arbitrary as far as this package is concerned.
+	key, _ := hex.DecodeString("2dd283abb148ebabcd894b306e3d86d0390c82a7")
+	err := c.Store(target, key, []string{"plz-out/gen/package/out1.txt"})
+	assert.NoError(t, err)
+	// Remove the old file, but remember its contents so we can compare later.
+	contents, err := ioutil.ReadFile("plz-out/gen/package/out1.txt")
+	assert.NoError(t, err)
+	err = os.Remove("plz-out/gen/package/out1.txt")
+	assert.NoError(t, err)
+	// Now retrieve back the output of this thing.
+	err = c.Retrieve(target, key)
+	assert.NoError(t, err)
+	cachedContents, err := ioutil.ReadFile("plz-out/gen/package/out1.txt")
+	assert.NoError(t, err)
+	assert.Equal(t, contents, cachedContents)
+}
+
+func newClient() *Client {
+	config := core.DefaultConfiguration()
+	config.Build.Path = []string{"/usr/local/bin", "/usr/bin", "/bin"}
+	// Can't use NewDefaultBuildState since we need to modify the config first.
+	state := core.NewBuildState(1, nil, 4, config)
```

Comment:
Here are those magic numbers again

Reply:
yup, same as above; as the comment mentions, NewDefaultBuildState would be simpler but we have to pass in the config first. Again, needs a wide-ranging refactor in its own PR.